File: app/src/main/java/com/absinthe/anywhere_/ui/shortcuts/ShortcutsActivity.java
package com.absinthe.anywhere_.ui.shortcuts;

import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.widget.ArrayAdapter;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.absinthe.anywhere_.AnywhereApplication;
import com.absinthe.anywhere_.R;
import com.absinthe.anywhere_.model.AnywhereEntity;
import com.absinthe.anywhere_.model.AnywhereType;
import com.absinthe.anywhere_.model.Const;
import com.absinthe.anywhere_.model.GlobalValues;
import com.absinthe.anywhere_.services.CollectorService;
import com.absinthe.anywhere_.utils.AppUtils;
import com.absinthe.anywhere_.utils.CommandUtils;
import com.absinthe.anywhere_.utils.PermissionUtils;
import com.absinthe.anywhere_.utils.TextUtils;
import com.absinthe.anywhere_.utils.UiUtils;
import com.absinthe.anywhere_.utils.handler.Opener;
import com.absinthe.anywhere_.utils.handler.URLSchemeHandler;
import com.absinthe.anywhere_.utils.manager.DialogManager;
import com.absinthe.anywhere_.utils.manager.Logger;
import com.absinthe.anywhere_.utils.manager.URLManager;
import com.absinthe.anywhere_.view.AnywhereDialogBuilder;
import com.absinthe.anywhere_.viewmodel.AnywhereViewModel;

import java.util.Objects;

public class ShortcutsActivity extends AppCompatActivity {
    public static final String ACTION_START_COLLECTOR = "START_COLLECTOR";
    public static final String ACTION_START_COMMAND = "START_COMMAND";
    public static final String ACTION_START_FROM_WIDGET = "START_FROM_WIDGET";
    public static final String ACTION_START_QR_CODE = "START_QR_CODE";
    public static final String ACTION_START_IMAGE = "START_IMAGE";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        UiUtils.setActionBarTransparent(this);
        AnywhereViewModel viewModel = new ViewModelProvider(this).get(AnywhereViewModel.class);
        Intent i = getIntent();
        String action = i.getAction();

        if (action != null) {
            if (action.equals(ACTION_START_COLLECTOR)) {
                if (GlobalValues.getWorkingMode().equals(Const.WORKING_MODE_URL_SCHEME)) {
                    AppUtils.openNewURLScheme(this);
                } else {
                    if (PermissionUtils.checkOverlayPermission(this, Const.REQUEST_CODE_ACTION_MANAGE_OVERLAY_PERMISSION)) {
                        CollectorService.startCollector(this);
                    }
                }
                finish();
            } else if (action.equals(ACTION_START_COMMAND)) {
                String cmd = i.getStringExtra(Const.INTENT_EXTRA_SHORTCUTS_CMD);
                if (cmd != null) {
                    if (cmd.startsWith(AnywhereType.DYNAMIC_PARAMS_PREFIX) ||
                            cmd.startsWith(AnywhereType.SHELL_PREFIX)) {
                        Opener.with(this)
                                .load(cmd)
                                .setOpenedListener(this::finish)
                                .open();
                    } else {
                        Opener.with(this).load(cmd).open();
                        finish();
                    }
                } else {
                    finish();
                }
            } else if (action.equals(ACTION_START_FROM_WIDGET)) {
                String cmd = i.getStringExtra(Const.INTENT_EXTRA_WIDGET_COMMAND);
                if (cmd != null) {
                    if (cmd.startsWith(AnywhereType.DYNAMIC_PARAMS_PREFIX) ||
                            cmd.startsWith(AnywhereType.SHELL_PREFIX)) {
                        Opener.with(this)
                                .load(cmd)
                                .setOpenedListener(this::finish)
                                .open();
                    } else {
                        Opener.with(this).load(cmd).open();
                        finish();
                    }
                } else {
                    finish();
                }
            } else if (action.equals(ACTION_START_QR_CODE)) {
                String id = i.getStringExtra(Const.INTENT_EXTRA_SHORTCUTS_CMD);
                CommandUtils.execCmd(id);
                finish();
            } else if (action.equals(Intent.ACTION_CREATE_SHORTCUT)) {
                viewModel.getAllAnywhereEntities().observe(this, anywhereEntities -> {
                    ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(AnywhereApplication.sContext, android.R.layout.select_dialog_singlechoice);
                    Logger.d("list=", anywhereEntities);
                    if (anywhereEntities != null) {
                        for (AnywhereEntity ae : anywhereEntities) {
                            arrayAdapter.add(ae.getAppName());
                        }
                    }
                    new AnywhereDialogBuilder(this)
                            .setAdapter(arrayAdapter, (dialogInterface, i1) -> {
                                Intent shortcutIntent = new Intent(ShortcutsActivity.this, ShortcutsActivity.class);
                                String cmd = TextUtils.getItemCommand(
                                        Objects.requireNonNull(anywhereEntities).get(i1));
                                if (cmd.startsWith(AnywhereType.QRCODE_PREFIX)) {
                                    shortcutIntent.setAction(ACTION_START_QR_CODE);
                                } else {
                                    shortcutIntent.setAction(ACTION_START_COMMAND);
                                }
                                shortcutIntent.putExtra(Const.INTENT_EXTRA_SHORTCUTS_CMD, cmd);

                                Intent intent = new Intent();
                                intent.putExtra(Intent.EXTRA_SHORTCUT_INTENT, shortcutIntent);
                                intent.putExtra(Intent.EXTRA_SHORTCUT_NAME, getString(R.string.shortcut_open));
                                intent.putExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE, Intent.ShortcutIconResource.fromContext(ShortcutsActivity.this, R.drawable.ic_shortcut_start_collector));
                                setResult(RESULT_OK, intent);
                                finish();
                            })
                            .setOnCancelListener(dialog -> finish())
                            .show();
                });
            } else if (action.equals(ACTION_START_IMAGE)) {
                String uri = i.getStringExtra(Const.INTENT_EXTRA_SHORTCUTS_CMD);
                if (uri != null) {
                    AnywhereEntity ae = AnywhereEntity.Builder();
                    ae.setParam1(uri);
                    DialogManager.showImageDialog(this, ae, this::finish);
                } else {
                    finish();
                }
            } else if (action.equals(Intent.ACTION_VIEW)) {
                Uri uri = i.getData();
                if (uri != null) {
                    String host = uri.getHost();

                    if (android.text.TextUtils.equals(host, URLManager.OPEN_HOST)) {
                        String param1 = uri.getQueryParameter(Const.INTENT_EXTRA_PARAM_1);
                        String param2 = uri.getQueryParameter(Const.INTENT_EXTRA_PARAM_2);
                        String param3 = uri.getQueryParameter(Const.INTENT_EXTRA_PARAM_3);

                        if (param1 != null && param2 != null && param3 != null) {
                            if (param2.isEmpty() && param3.isEmpty()) {
                                try {
                                    URLSchemeHandler.parse(param1, this);
                                } catch (ActivityNotFoundException e) {
                                    e.printStackTrace();
                                }
                            } else {
                                AnywhereEntity ae = AnywhereEntity.Builder();
                                ae.setParam1(param1);
                                ae.setParam2(param2);
                                ae.setParam3(param3);
                                ae.setType(AnywhereType.ACTIVITY);

                                CommandUtils.execCmd(TextUtils.getItemCommand(ae));
                            }
                        }
                    }
                }
                finish();
            }
        } else {
            finish();
        }
    }
}

File: app/src/main/java/com/absinthe/anywhere_/ui/shortcuts/ShortcutsActivity.kt
package com.absinthe.anywhere_.ui.shortcuts

import android.app.Activity
import android.content.*
import android.os.Bundle
import android.os.IBinder
import android.widget.ArrayAdapter
import androidx.activity.viewModels
import androidx.viewbinding.ViewBinding
import com.absinthe.anywhere_.BaseActivity
import com.absinthe.anywhere_.R
import com.absinthe.anywhere_.constants.AnywhereType
import com.absinthe.anywhere_.constants.Const
import com.absinthe.anywhere_.constants.EventTag
import com.absinthe.anywhere_.constants.GlobalValues
import com.absinthe.anywhere_.model.ExtraBean
import com.absinthe.anywhere_.model.database.AnywhereEntity
import com.absinthe.anywhere_.services.overlay.CollectorService
import com.absinthe.anywhere_.services.overlay.ICollectorService
import com.absinthe.anywhere_.services.tile.TILE_ACTIVE_STATE
import com.absinthe.anywhere_.ui.editor.impl.SWITCH_OFF
import com.absinthe.anywhere_.utils.AppUtils.openNewURLScheme
import com.absinthe.anywhere_.utils.CommandUtils
import com.absinthe.anywhere_.utils.ToastUtil
import com.absinthe.anywhere_.utils.UxUtils
import com.absinthe.anywhere_.utils.handler.Opener
import com.absinthe.anywhere_.utils.manager.DialogManager.showImageDialog
import com.absinthe.anywhere_.utils.manager.URLManager
import com.absinthe.anywhere_.view.app.AnywhereDialogBuilder
import com.absinthe.anywhere_.view.app.AnywhereDialogFragment
import com.absinthe.anywhere_.viewmodel.AnywhereViewModel
import com.blankj.utilcode.util.Utils
import com.google.gson.Gson
import com.microsoft.appcenter.analytics.Analytics
import timber.log.Timber

class ShortcutsActivity : BaseActivity<ViewBinding>() {

    private var isBound = false
    private var collectorService: ICollectorService? = null

    private val viewModel by viewModels<AnywhereViewModel>()

    override fun setViewBinding(): Nothing? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        UxUtils.setActionBarTransparent(this)
        Analytics.trackEvent(EventTag.SHORTCUT_OPEN)
        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        intent.action?.let {
            Timber.d("action = %s", it)

            when (it) {
                ACTION_START_COLLECTOR -> {
                    if (GlobalValues.workingMode == Const.WORKING_MODE_URL_SCHEME) {
                        openNewURLScheme(this)
                    } else {
                        if (isBound) {
                            collectorService?.startCollector()
                        } else {
                          CollectorService.serviceConnection = object : ServiceConnection {
                            override fun onServiceDisconnected(name: ComponentName?) {
                              isBound = false
                              collectorService = null
                            }

                            override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
                              isBound = true
                              collectorService = ICollectorService.Stub.asInterface(service)
                              collectorService?.startCollector()
                            }

                          }
                            applicationContext.bindService(Intent(this, CollectorService::class.java), CollectorService.serviceConnection!!, Context.BIND_AUTO_CREATE)
                        }
                    }
                    shouldFinishOnResume = true
                }
                ACTION_START_ENTITY -> {
                    intent.getStringExtra(Const.INTENT_EXTRA_SHORTCUTS_ID)?.let { id ->
                        viewModel.allAnywhereEntities.observe(this) { list ->
                          list.find { findItem ->
                            findItem.id == id
                          }?.apply {
                            Opener.with(this@ShortcutsActivity)
                              .load(this)
                              .setOpenedListener(object : Opener.OnOpenListener {
                                override fun onOpened() {
                                  shouldFinishOnResume = true
                                  finish()
                                }
                              })
                              .open()

                            val tileExtra = intent.getStringExtra(Const.INTENT_EXTRA_FROM_TILE)
                            if (tileExtra != null && this.type == AnywhereType.Card.SWITCH_SHELL) {
                              GlobalValues.mmkv.encode(
                                tileExtra + TILE_ACTIVE_STATE,
                                this.param3 != SWITCH_OFF
                              )
                            }
                          }
                        }
                    }
                }
                ACTION_START_FROM_WIDGET -> {
                    intent.getParcelableExtra<AnywhereEntity>(Const.INTENT_EXTRA_WIDGET_ENTITY)?.let { entity ->
                        Opener.with(this@ShortcutsActivity)
                            .load(entity)
                            .setOpenedListener(object : Opener.OnOpenListener {
                                override fun onOpened() {
                                    shouldFinishOnResume = true
                                }
                            })
                            .open()
                    } ?: let { shouldFinishOnResume = true }
                }
                Intent.ACTION_CREATE_SHORTCUT -> {
                    viewModel.allAnywhereEntities.observe(this) { anywhereEntities: List<AnywhereEntity>? ->
                      val arrayAdapter = ArrayAdapter<String>(
                        Utils.getApp(),
                        android.R.layout.select_dialog_singlechoice
                      )

                      anywhereEntities?.let { entities ->
                        Timber.d("list = %s", entities)

                        for (ae in entities) {
                          arrayAdapter.add(ae.appName)
                        }

                        AnywhereDialogBuilder(this)
                          .setAdapter(arrayAdapter) { _: DialogInterface?, i: Int ->
                            val entity = entities[i]
                            val shortcutIntent =
                              Intent(this@ShortcutsActivity, ShortcutsActivity::class.java).apply {
                                if (entities[i].type == AnywhereType.Card.IMAGE) {
                                  action = ACTION_START_IMAGE
                                  putExtra(Const.INTENT_EXTRA_SHORTCUTS_CMD, entity.param1)
                                } else {
                                  action = ACTION_START_ENTITY
                                  putExtra(Const.INTENT_EXTRA_SHORTCUTS_ID, entity.id)
                                }
                              }

                            setResult(Activity.RESULT_OK, Intent().apply {
                              putExtra(Intent.EXTRA_SHORTCUT_INTENT, shortcutIntent)
                              putExtra(
                                Intent.EXTRA_SHORTCUT_NAME,
                                getString(R.string.shortcut_open)
                              )
                              putExtra(
                                Intent.EXTRA_SHORTCUT_ICON_RESOURCE,
                                Intent.ShortcutIconResource.fromContext(
                                  this@ShortcutsActivity, R.drawable.ic_shortcut_start_collector
                                )
                              )
                            })
                            shouldFinishOnResume = true
                          }
                          .setOnCancelListener {
                            shouldFinishOnResume = true
                            finish()
                          }
                          .show()
                      }
                    }
                }
                ACTION_START_IMAGE -> {
                    intent.getStringExtra(Const.INTENT_EXTRA_SHORTCUTS_CMD)?.let { uri ->
                        showImageDialog(
                            this,
                            uri,
                            object : AnywhereDialogFragment.OnDismissListener {
                                override fun onDismiss() {
                                    shouldFinishOnResume = true
                                    finish()
                                }
                            })
                    } ?: run { shouldFinishOnResume = true }
                }
                ACTION_START_DEVICE_CONTROL -> {
                    val type = intent.getIntExtra(Const.INTENT_EXTRA_TYPE, -1)
                    val param1 = intent.getStringExtra(Const.INTENT_EXTRA_PARAM_1) ?: return@let
                    val param2 = intent.getStringExtra(Const.INTENT_EXTRA_PARAM_2) ?: return@let
                    val param3 = intent.getStringExtra(Const.INTENT_EXTRA_PARAM_3) ?: return@let
                    val entity = AnywhereEntity().apply {
                        this.type = type
                        this.param1 = param1
                        this.param2 = param2
                        this.param3 = param3
                    }
                    Opener.with(this@ShortcutsActivity)
                            .load(entity)
                            .setOpenedListener(object : Opener.OnOpenListener {
                                override fun onOpened() {
                                    shouldFinishOnResume = true
                                }
                            })
                            .open()
                }
                Intent.ACTION_VIEW -> {
                    intent.data?.let { uri ->
                        if (uri.host == URLManager.OPEN_HOST) {
                            var dynamicParam: ExtraBean.ExtraItem? = null
                            uri.getQueryParameter(Const.INTENT_EXTRA_DYNAMIC_PARAM)
                                ?.let { dynamic ->
                                    try {
                                        dynamicParam = Gson().fromJson(
                                            dynamic,
                                            ExtraBean.ExtraItem::class.java
                                        )
                                    } catch (ignore: Exception) {
                                    }
                                }
                          var dynamicParams: Array<ExtraBean.ExtraItem>? = null
                          uri.getQueryParameter(Const.INTENT_EXTRA_DYNAMICS_PARAM)
                            ?.let { dynamics ->
                              try {
                                dynamicParams = Gson().fromJson(
                                  dynamics,
                                  Array<ExtraBean.ExtraItem>::class.java
                                )
                              } catch (ignore: Exception) {
                              }
                            }
                            uri.getQueryParameter(Const.INTENT_EXTRA_OPEN_SHORT_ID)?.let { sid ->
                                viewModel.allAnywhereEntities.observe(this) { list ->
                                  list.find { findItem ->
                                    findItem.id.endsWith(sid)
                                  }?.apply {
                                    Opener.with(this@ShortcutsActivity)
                                      .load(this)
                                      .setDynamicExtra(dynamicParam)
                                      .setDynamicExtras(dynamicParams)
                                      .setOpenedListener(object : Opener.OnOpenListener {
                                        override fun onOpened() {
                                          shouldFinishOnResume = true
                                        }
                                      })
                                      .open()
                                  } ?: run {
                                    ToastUtil.makeText(R.string.toast_invaild_sid)
                                    shouldFinishOnResume = true
                                  }
                                }
                            } ?: run { shouldFinishOnResume = true }
                        }
                    }
                }
                else -> shouldFinishOnResume = true
            }
        }
        if (intent.getBooleanExtra(Const.INTENT_EXTRA_EMULATE_BACK_PRESS, false)) {
            CommandUtils.execAdbCmd(Const.CMD_BACK_PRESS)
        }
    }

    companion object {
        const val ACTION_START_COLLECTOR = "START_COLLECTOR"
        const val ACTION_START_ENTITY = "START_ENTITY"
        const val ACTION_START_FROM_WIDGET = "START_FROM_WIDGET"
        const val ACTION_START_IMAGE = "START_IMAGE" //Old Scheme
        const val ACTION_START_DEVICE_CONTROL = "ACTION_START_DEVICE_CONTROL"
    }
}
